cloud/openstack/os_security_group_rule.py

```diff
@@ -0,0 +1,257 @@
+      description:
+        - Starting port
+      required: true
+   port_range_max:
+      description:
+        - Ending port
+      required: true
+   remote_ip_prefix:
+      description:
+        - Source IP address(es) in CIDR notation (exclusive with remote_group)
+      required: false
+   remote_group:
+      description:
+        - ID of Security group to link (exclusive with remote_ip_prefix)
+      required: false
+   ethertype:
+      description:
+        - Must be IPv4 or IPv6, and addresses represented in CIDR must
+          match the ingress or egress rules. Not all providers support IPv6.
+      choices: ['IPv4', 'IPv6']
+      default: IPv4
+   direction:
+     description:
+        - The direction in which the security group rule is applied. Not
+          all providers support egress.
+     choices: ['egress', 'ingress']
+     default: ingress
+   state:
+     description:
+       - Should the resource be present or absent.
+     choices: [present, absent]
+     default: present
+requirements: ["shade"]
+'''
+# TODO(mordred): add ethertype and direction
+
+EXAMPLES = '''
+# Create a security group rule
+- os_security_group_rule:
+    cloud: mordred
+    security_group: foo
+    protocol: tcp
+    port_range_min: 80
+    port_range_max: 80
+    remote_ip_prefix: 0.0.0.0/0
+'''
+
+RETURN = '''
+id:
+  description: Unique rule UUID.
+  type: string
+direction:
+  description: The direction in which the security group rule is applied.
+  type: string
+  sample: 'egress'
+ethertype:
+  description: One of IPv4 or IPv6.
+  type: string
+  sample: 'IPv4'
+port_range_min:
+  description: The minimum port number in the range that is matched by
+               the security group rule.
+  type: int
+  sample: 8000
+port_range_max:
+  description: The maximum port number in the range that is matched by
+               the security group rule.
+  type: int
+  sample: 8000
+protocol:
+  description: The protocol that is matched by the security group rule.
+  type: string
+  sample: 'tcp'
+remote_ip_prefix:
+  description: The remote IP prefix to be associated with this security group rule.
+  type: string
+  sample: '0.0.0.0/0'
+security_group_id:
+  description: The security group ID to associate with this security group rule.
+  type: string
+'''
+
+
+def _find_matching_rule(module, secgroup):
+    """
+    Find a rule in the group that matches the module parameters.
+    :returns: The matching rule dict, or None if no matches.
+    """
+    protocol = module.params['protocol']
+    port_range_min = module.params['port_range_min']
+    port_range_max = module.params['port_range_max']
+    remote_ip_prefix = module.params['remote_ip_prefix']
+    ethertype = module.params['ethertype']
+    direction = module.params['direction']
+
+    for rule in secgroup['security_group_rules']:
+        # No port, or -1, will be returned from shade as None
+        if rule['port_range_min'] is None:
+            rule_port_range_min = -1
```

Comment:
@j2sol Yep, you were correct. It was easier to just convert incoming -1 values to None so that they match what is returned from shade. I added a lot of inline documentation to describe what the various actors are doing and how we need to do comparisons. I also created a playbook to test *several* various scenarios, and tested against both Nova and Neutron, and it looks pretty solid now. Hopefully another set of eyes will help catch anything I may have missed.